### PR TITLE
Add Form Builder API Implementation

### PR DESCRIPTION
**Pull request by Gravatronics**
See #2509

### CODE REVIEW

File: com.woltlab.wcf/templates/__iconFormField.tpl [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+		var button = elById('{@$field->getPrefixedId()}_openIconDialog');
+		var icon = elById('{@$field->getPrefixedId()}_icon');
+		var input = elById('{@$field->getPrefixedId()}');
+		
+		var callback = function(iconName) {
+			icon.className = 'icon icon64 fa-' + iconName;
+			input.value = iconName;
+		};
+		
+		button.addEventListener('click', function() {
```
Comment (joshuaruesweg) [inline com.woltlab.wcf/templates/__iconFormField.tpl]: `button.addEventListener('click', function() { ` => `button.addEventListener(WCF_CLICK_EVENT, function() {`

File: wcfsetup/install/files/acp/templates/__iconFormField.tpl [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+		var button = elById('{@$field->getPrefixedId()}_openIconDialog');
+		var icon = elById('{@$field->getPrefixedId()}_icon');
+		var input = elById('{@$field->getPrefixedId()}');
+		
+		var callback = function(iconName) {
+			icon.className = 'icon icon64 fa-' + iconName;
+			input.value = iconName;
+		};
+		
+		button.addEventListener('click', function() {
```
Comment (joshuaruesweg) [inline wcfsetup/install/files/acp/templates/__iconFormField.tpl]: `button.addEventListener('click', function() { ` => `button.addEventListener(WCF_CLICK_EVENT, function() {`

File: wcfsetup/install/files/lib/system/form/builder/IFormParentNode.class.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	 * 
+	 * @param	string		$nodeId		id of searched node
+	 * @return	bool
+	 */
+	public function contains($nodeId);
+	
+	/**
+	 * Returns a recursive iterator for this node.
+	 * 
+	 * Note: A class cannot implement `\Iterator` and `\IteratorAggregate` at the same time.
```
[the remark this answers is not in the document]
Comment (Gravatronics) [reply wcfsetup/install/files/lib/system/form/builder/IFormParentNode.class.php]: You can iterate over the direct children of the element or iterate recursively over the complete tree with this object being the root.

File: wcfsetup/install/files/lib/form/AbstractFormBuilderForm.class.php [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	public function readData() {
+		if ($this->formObject !== null) {
+			$this->form->loadValuesFromObject($this->formObject);
+		}
+		else if ($this->formAction === 'edit') {
```
Comment (Fabii547) [inline wcfsetup/install/files/lib/form/AbstractFormBuilderForm.class.php]: `'edit'` -> `'update'`?

File: wcfsetup/install/files/lib/system/form/builder/IFormParentNode.class.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimWolla) [reply wcfsetup/install/files/lib/system/form/builder/IFormParentNode.class.php]: I feel this method should be called `getRecursiveIterator` then to make the difference clear at call time.

**Pull request merged**